src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -261,27 +375,160 @@ private List<Index> GetImpactedIndexesToWaitForToBecomeNonStale(HashSet<string>
+
+            protected TranscationMergedCommand(DocumentDatabase database)
+            {
+                Database = database;
+            }
+
+            protected void AddPutResult(DocumentsStorage.PutOperationResults putResult)
+            {
+                LastChangeVector = putResult.ChangeVector;
+                ModifiedCollections?.Add(putResult.Collection.Name);
+
+                // Make sure all the metadata fields are always been add
+                var putReply = new DynamicJsonValue
+                {
+                    ["Type"] = nameof(CommandType.PUT),
+                    [Constants.Documents.Metadata.Id] = putResult.Id,
+                    [Constants.Documents.Metadata.Collection] = putResult.Collection.Name,
+                    [Constants.Documents.Metadata.ChangeVector] = putResult.ChangeVector,
+                    [Constants.Documents.Metadata.LastModified] = putResult.LastModified
+                };
+
+                if (putResult.Flags != DocumentFlags.None)
+                    putReply[Constants.Documents.Metadata.Flags] = putResult.Flags;
+
+                Reply.Add(putReply);
+            }
+
+            protected void AddDeleteResult(DocumentsStorage.DeleteOperationResult? deleted, string id)
+            {
+                if (deleted != null)
+                {
+                    LastTombstoneEtag = deleted.Value.Etag;
+                    ModifiedCollections?.Add(deleted.Value.Collection.Name);
+                }
+
+                Reply.Add(new DynamicJsonValue
+                {
+                    [nameof(BatchRequestParser.CommandData.Id)] = id,
+                    [nameof(BatchRequestParser.CommandData.Type)] = nameof(CommandType.DELETE),
+                    ["Deleted"] = deleted != null
+                });
+            }
+        }
+
+        public class ClusterTransactionMergedCommand : TranscationMergedCommand
+        {
+            private readonly TransactionOperationContext _serverContext;
+            private readonly long _index;
+            public bool Skipped { get; private set; }
+            public ClusterTransactionCommand.ClusterTransactionOptions Options { get; private set; }
+
+            public ClusterTransactionMergedCommand(DocumentDatabase database, TransactionOperationContext serverContext, long index) : base(database)
+            {
+                _serverContext = serverContext;
+                _index = index;
+            }
+
+            public override int Execute(DocumentsOperationContext context)
+            {
+                var global = DocumentsStorage.GetDatabaseChangeVector(context);
+                var clusterId = Database.ServerStore.Engine.ClusterBase64Id;
+                var currentRaftIndex = ChangeVectorUtils.GetEtagById(global, clusterId);
+
+                // Check if we already have the documents of this cluster transaction.
+                if (currentRaftIndex >= _index)
+                {
+                    Skipped = true;
```

Comment:
Okay, this isn't correct. But I'm not sure what should be done here.
The problem here is that you can't really return the current value, because it doesn't reflect the command, it might have already been updated again

Reply:
Still wrong